### PR TITLE
Build: implementation of `build.commands`

### PR DESCRIPTION
**Pull request by humitos**
Initial implementation of `build.commands` as a beta feature to expose to user and collect more use cases.

## Build without using any tool

```yaml
version: 2

build:
  os: ubuntu-20.04
  commands:
    - mkdir output/
    - echo "Hello world! `date`" > output/index.html

  tools:  # <--- this shouldn't be required for these cases where no tools are needed for the build
    python: "3"
```

(https://github.com/readthedocs/test-builds/blob/build-commands/.readthedocs.yaml)

### Build details page

Screenshot_2022-04-27_18-13-06

### `index.html` file served by El Proxito

Screenshot_2022-04-27_18-13-16

## Build using `pelican`

```yaml
version: 2

build:
  os: ubuntu-22.04
  tools:
    python: "3.10"
  commands:
    - pip install pelican[markdown]
    # After installing Python packages with binaries on it, we need to call `asdf reshim`
    # (recreate shims for version of a package)
    # https://github.com/danhper/asdf-python/issues/37
    - asdf reshim
    - pelican --settings docs/pelicanconf.py --output output/ docs/
```

(https://github.com/readthedocs/test-builds/blob/pelican/.readthedocs.yaml)

### Build details page

Screenshot_2022-05-23_11-59-06

### `index.html` file served by El Proxito

Screenshot_2022-05-23_11-50-31

----

## Decisions / Notes

- Expose the feature as beta, communicating this in the documentation and also on the build details page when we detect `build.commands` on the config file
- It defines an "implicit contract" where the `output/` folder under the repository checkout will be uploaded to S3
- No HTML integrations will be done (no flyout, no ads, no search)
- No extra formats like ePUB, PDF and HTMLZip will be available
- Without the flyout integration, this is probably more useful for single-version projects
- `build.os` and `build.tools` are mandatory when using `build.commands`
- We should probably make `build.tools` _not required_ since it's possible to generate HTML without any of these languages as shown in the first example where only `echo` command is used
- `asdf` requires re-shim the binaries after installing new CLI tools with Python to be able to find them when executing any of these commands. There was an attempt to re-shim automatically at https://github.com/danhper/asdf-python/issues/37 but they removed the `post_<tool>_<shim>` because https://github.com/asdf-vm/asdf/pull/502. This may need some extra research to make the UX better. However, for now, I'm adding `asdf reshim` to my `build.commands`. We could execute `asdf reshim python` after each of the `build.commands`, for now as well
- ~~I didn't update the "Build customization" page, but I'm happy to add a small example there (https://docs.readthedocs.io/en/stable/build-customization.html) if we consider it's good at this point. However, it may require a good amount of re-writing on that page to make the distinction between the two approaches clear~~

Related #9062 and others linked into that issue.

### CODE REVIEW

**Comment by ericholscher** [comment]
Neat how little code this is. Definitely needs a lot more done around the edges to make it nicer, but great that we already have all the basic infrastructure to make it work. 👍

File: readthedocs/projects/tasks/builds.py [hunk 1]
```diff
@@ -595,6 +596,16 @@ def execute(self):
         # ``__exit__``
         self.data.build_director.create_build_environment()
         with self.data.build_director.build_environment:
+
+            # NOTE: check if the build uses `build.commands` and only run those
+            if self.data.config.build.commands:
+                self.update_build(state=BUILD_STATE_BUILDING)
```
Comment (ericholscher) [inline readthedocs/projects/tasks/builds.py:602]: I think we're 👍  on moving forward with this in some kind of beta feature. Either a feature flag, or some very large warning in the docs that this beta and isn't supported yet.

Comment (humitos) [reply readthedocs/projects/tasks/builds.py:602]: I'm fine moving forward with this to gather some feedback.

Regarding how to communicate,

- I think using a feature flag may generate some extra work in our support queue that I'd like to avoid if possible
- Also, with a feature flag, fewer people will test this because the entry barrier is higher
- I like the idea of adding a warning in the docs communicating this is beta and is not supported yet, but I think we need something else on the build page as well
- I'd add a warning message on each build details page that uses `build.commands` too, to communicate this effectively to users that copy and paste the config from other projects without reading our docs 😄 

I'm 👍🏼 to add this without a feature flag + in beta status + warning in docs + warning in build page.

**Comment by humitos** [comment]
I'm asking review for the three of you @ericholscher, @agjohnson, and @stsewd because I think this is an important change that will mold our path moving forward regarding the UX.

File: readthedocs/doc_builder/director.py [hunk 2]
```diff
@@ -335,6 +339,163 @@ def run_build_job(self, job):
...
+    def run_build_commands(self):
+        cwd = self.data.project.checkout_path(self.data.version.slug)
+        environment = self.vcs_environment
+        for command in self.data.config.build.commands:
+            environment.run(*command.split(), escape_command=False, cwd=cwd)
+
+        # Copy files to artifacts path so they are uploaded to S3
+        target = self.data.project.artifact_path(
+            version=self.data.version.slug,
+            type_="sphinx",
```
Comment (stsewd) [inline readthedocs/doc_builder/director.py:351]: I think we should introduce a new doctype, the doctype is used in several places to make some decisions (specially when generating links, foo/ vs /foo.html). It should also make it easier to filter projects/versions using this new feature.

File: readthedocs/doc_builder/python_environments.py [hunk 3]
```diff
@@ -40,138 +37,6 @@ def __init__(self, version, build_env, config=None):
             version_slug=self.version.slug,
         )
 
-    def install_build_tools(self):
```
[the remark this answers is not in the document]
Comment (humitos) [reply readthedocs/doc_builder/python_environments.py:43]: I moved this because it's agnostic to the language. It could install only NodeJS. So, it doesn't make sense to be on `PythonEnvironment` class. I needed to call `install_build_tools` _without_ creating an environment when using `build.commands`.

File: readthedocs/doc_builder/director.py (hunk 2 above)
Comment (humitos) [reply readthedocs/doc_builder/director.py:351]: I'm not aware of how these `type_`s are used. If we use `generic` here, what other changes in logic do we need to do? Will it touch multiple places?

File: docs/user/build-customization.rst [hunk 4]
```diff
@@ -218,3 +223,36 @@ To setup it, you need to define the version of Node.js to use and install the de
          - npm ci
          # Install any other extra dependencies to build the docs
          - npm install -g jsdoc
+
+
+Override the build process completely
```
Comment (ericholscher) [inline docs/user/build-customization.rst:228]: This is going to be hidden under all the examples. I'd argue this is a different page.

File: docs/user/build-customization.rst [hunk 5]
```diff
@@ -218,3 +223,36 @@ To setup it, you need to define the version of Node.js to use and install the de
...
+     tools:
+       python: "3.10"
+     commands:
+       - pip install pelican[markdown]
+       - pelican --settings docs/pelicanconf.py --output output/ docs/
+
+
+As Read the Docs does not have control over the build process,
+you are responsible to run all the commands required to install the requirements and build the documentation properly.
+Once the build process finishes, the ``output/`` folder will be hosted.
```
Comment (ericholscher) [inline docs/user/build-customization.rst:258]: This should be much more obvious in this page. This feels hidden currently. Preferably in bold, and under a heading showing you should _definitely_ read it.

File: docs/user/config-file/v2.rst [hunk 6]
```diff
@@ -456,6 +456,43 @@ See :doc:`/build-customization` for more details.
...
+     os: ubuntu-22.04
+     tools:
+       python: "3.10"
+     commands:
+       - pip install pelican
+       - pelican --settings docs/pelicanconf.py --output output/ docs/
+
+.. note::
+
+   ``build.os`` and ``build.tools`` are also required when using ``build.commands``.
```
Comment (ericholscher) [inline docs/user/config-file/v2.rst:489]: Yea, feels like build.tools being required here is a bit odd? If I'm using a Hugo, I don't need any of our supported tools..

File: readthedocs/doc_builder/director.py [hunk 7]
```diff
@@ -335,6 +339,163 @@ def run_build_job(self, job):
...
+        )
+        artifacts_path = os.path.join(cwd, "output")
```
Comment (ericholscher) [inline readthedocs/doc_builder/director.py:353]: This should probably be a constant, at least.

File: readthedocs/doc_builder/director.py [hunk 8]
```diff
@@ -335,6 +339,163 @@ def run_build_job(self, job):
...
+
+        shutil.copytree(
+            artifacts_path,
+            target,
+        )
+
+        self.data.outcomes = defaultdict(lambda: False)
+        self.data.outcomes["html"] = True
+
+    def install_build_tools(self):
```
Comment (ericholscher) [inline readthedocs/doc_builder/director.py:365]: Would be useful to note in a comment that this was moved & not changed, since it's hard to know if it was changed when reviewing. I was also surprised that it was here, but then realized it was a refactor.

File: readthedocs/doc_builder/exceptions.py [hunk 9]
```diff
@@ -36,6 +36,9 @@ class BuildUserError(BuildBaseException):
         "To resolve this error, double check your project configuration and installed "
         "dependencies are correct and have not changed recently."
     )
+    BUILD_COMMANDS_WITHOUT_OUTPUT = gettext_noop(
+        'No "output/" folder detected on the root of the repository after building the project.'
```
Comment (ericholscher) [inline readthedocs/doc_builder/exceptions.py:40]: I don't love this error message (`root of the repository` seems hard to parse), but I don't have a great suggestion for improving it..

Maybe:

```suggestion
        'No "output/" folder was created in the root directory after running `build.commands`.'
```

File: readthedocs/templates/builds/build_detail.html [hunk 10]
```diff
@@ -175,6 +175,17 @@
       {% endif %}
     {% endif %}
 
+    {% if build.finished and build.config.build.commands %}
+    <div class="build-ideas">
+      <p>
+        {% blocktrans trimmed with config_file_link="https://docs.readthedocs.io/page/build-customization.html#override-the-build-process-completely" %}
+        <strong><code>"build.commands"</code> config is in beta, use it under your own risks.</strong>
```
Comment (ericholscher) [inline readthedocs/templates/builds/build_detail.html:182]: 👍 
```suggestion
        <strong>The <code>"build.commands"</code> feature is in beta, and could have backwards incompatible changes while in beta.</strong>
```

File: docs/user/build-customization.rst [hunk 11]
```diff
@@ -218,3 +223,36 @@ To setup it, you need to define the version of Node.js to use and install the de
...
+   However, integrating all of them is part of the plan.
+   Use it under your own responsibility.
+
+
+When :ref:`extending the build process <build-customization:extend the build process>` is not enough for a particular use case,
+all the default commands can be overwritten to execute only the commands defined by the project on its config file.
+This can be achieve by using :ref:`config-file/v2:build.commands` key as shown in the following example building a `Pelican <https://blog.getpelican.com/>`_ site:
+
+
+.. code:: yaml
```
Comment (ericholscher) [inline docs/user/build-customization.rst:244]: I feel like we want a minimal example here as well. Perhaps the `echo` example in the PR description, so people understand how it works without knowing pelican.

File: docs/user/config-file/v2.rst [hunk 12]
```diff
@@ -456,6 +456,43 @@ See :doc:`/build-customization` for more details.
...
 
 
+build.commands
+``````````````
+
+Specify a list of commands that Read the Docs will run on the build process.
+When ``build.commands`` is used, none of the :term:`pre-defined build jobs` will be executed.
+(see :doc:`/build-customization` for more details).
+This allows you to run custom commands and control the build process completely.
+The ``output/`` directory (relative to the checkout's path) will be uploaded and hosted by Read the Docs.
```
Comment (ericholscher) [inline docs/user/config-file/v2.rst:466]: I do wonder if we want to use something like `output/html/`, so that we have a more obvious path towards supporting other types? We could always just `output_pdf` in the future, but that will be weird that it isn't `output_html`..

**Review by ericholscher — COMMENTED** [review]
This looks like a great start! I think there's a balance here between doing too much work to implement this, but also think enough ahead to not break things without a good reason. 

> We could execute asdf reshim python after each of the build.commands, for now as well

This seems like something we should avoid. The whole goal here is to remove magic, so let's not add it back in..

File: docs/user/build-customization.rst [hunk 13]
```diff
@@ -218,3 +223,36 @@ To setup it, you need to define the version of Node.js to use and install the de
...
+
+   This feature is in a *beta phase* and could suffer incompatible changes or even removed completely in the near feature.
+   It does not yet support some of the Read the Docs' integrations like the :term:`flyout menu`, search and ads.
+   However, integrating all of them is part of the plan.
+   Use it under your own responsibility.
+
+
+When :ref:`extending the build process <build-customization:extend the build process>` is not enough for a particular use case,
+all the default commands can be overwritten to execute only the commands defined by the project on its config file.
+This can be achieve by using :ref:`config-file/v2:build.commands` key as shown in the following example building a `Pelican <https://blog.getpelican.com/>`_ site:
```
Comment (agjohnson) [inline docs/user/build-customization.rst:241]: I feel like real-life examples of *why* you might use this feature are really helpful in explaining this feature. I do like the Pelican example for that.

That is, "if you need more control" only vaguely describes why a user might use this feature. We might want to be more clear that it's "if your project uses a different documentation tool".

But just to clarify and reduce for now:

```suggestion
If your project requires more control of the build process,
all the commands executed during builds can be overridden using the :ref:`config-file/v2:build.commands` configuration file key.

For example, if your project uses `Pelican <https://blog.getpelican.com/>`_ instead of Sphinx for it's documentation, your project could use the following configuration file:
```

File: docs/user/build-customization.rst (hunk 11 above)
Comment (agjohnson) [reply docs/user/build-customization.rst:244]: I actually like the Pelican example, but perhaps we could use a more common documentation tool. What I like about this example is it relates the usage to the user -- use `build.commands` if you are using a documentation tool we don't support.

Comment (agjohnson) [inline docs/user/build-customization.rst:244]: Also:

```suggestion
.. code:: yaml
    :caption: .readthedocs.yaml
```

File: docs/user/build-customization.rst [hunk 14]
```diff
@@ -218,3 +223,36 @@ To setup it, you need to define the version of Node.js to use and install the de
...
+     os: "ubuntu-22.04"
+     tools:
+       python: "3.10"
+     commands:
+       - pip install pelican[markdown]
+       - pelican --settings docs/pelicanconf.py --output output/ docs/
+
+
+As Read the Docs does not have control over the build process,
+you are responsible to run all the commands required to install the requirements and build the documentation properly.
```
Comment (agjohnson) [inline docs/user/build-customization.rst:257]:
```suggestion
you are responsible for running all the commands required to install requirements and build the documentation properly.
```

File: docs/user/build-customization.rst (hunk 5 above)
Comment (agjohnson) [reply docs/user/build-customization.rst:258]: It does feel like we're missing some of the requirements here. I feel like we need a list of things. It would help to define the current working directory (repo checkout path), the output path that is synced (`output/`), and might be worth mentioning that `asdf` is used underneath everything too (you probably need to `asdf reshim`)

File: docs/user/config-file/v2.rst (hunk 12 above)
Comment (agjohnson) [reply docs/user/config-file/v2.rst:466]: I had this same feedback. It would be great to have an `output/{html,pdf,epub}/` from the start.

File: docs/user/config-file/v2.rst [hunk 15]
```diff
@@ -456,6 +456,43 @@ See :doc:`/build-customization` for more details.
...
+
+.. warning::
+
+   This feature is in a *beta phase* and could suffer incompatible changes or even removed completely in the near feature.
+   It does not yet support some of the Read the Docs' integrations like the :term:`flyout menu`, search and ads.
+   However, integrating all of them is part of the plan.
+   Use it under your own responsibility.
```
Comment (agjohnson) [inline docs/user/config-file/v2.rst:473]: I had edits on this same copy above.

File: docs/user/config-file/v2.rst [hunk 16]
```diff
@@ -456,6 +456,43 @@ See :doc:`/build-customization` for more details.
...
+
+.. code-block:: yaml
```
Comment (agjohnson) [inline docs/user/config-file/v2.rst:475]:
```suggestion
.. code-block:: yaml
    :caption: .readthedocs.yaml
```

File: docs/user/config-file/v2.rst (hunk 6 above)
Comment (agjohnson) [reply docs/user/config-file/v2.rst:489]: I was confused by this with `build.jobs` as well.

File: readthedocs/doc_builder/director.py (hunk 8 above)
Comment (agjohnson) [reply readthedocs/doc_builder/director.py:365]: I am a strong +1 on separating moving files, big linting, and similar operations into independent commits. Review is much easier if moving code is isolated from changing code. I'd even be +1 on setting a policy on this, as it does take a good deal of energy from reviewers.

File: readthedocs/doc_builder/exceptions.py (hunk 9 above)
Comment (agjohnson) [reply readthedocs/doc_builder/exceptions.py:40]: We could say "checkout path", but even that's probably not clear. "root directory" implies `/` to me though.

We could not mention the path in the error, but instead do a better job of highlighting requirements in the documentation. If there was something obvious to reference in our docs, that feels most helpful to the user. Nomenclature here will be difficult to pin down.

File: readthedocs/templates/builds/build_detail.html [hunk 17]
```diff
@@ -175,6 +175,17 @@
       {% endif %}
     {% endif %}
 
+    {% if build.finished and build.config.build.commands %}
```
Comment (agjohnson) [inline readthedocs/templates/builds/build_detail.html:178]: In general, I want to do away with these types of blocks in the build template. This is where I've described wanting code constructs to link error/warnings. But, that's a conversation we've been having for a while now.

But specifically to this block, this feels like an odd place to warn the user. The user has already enabled the feature for their builds for this to block to show. Why is it important to warn the user again at this point?

Perhaps the intention of this block was to instead say that "build.commands is in use for this build. If your build is hosed, you're on your own."

File: readthedocs/doc_builder/director.py (hunk 2 above)
Comment (stsewd) [reply readthedocs/doc_builder/director.py:351]: The more prominent places where this used are

- version.documentation_type
   - https://github.com/readthedocs/readthedocs.org/blob/d3351d52c41507f4c5f1ba737f8ef0b684d6e642/readthedocs/api/v2/views/footer_views.py#L158-L158
   - https://github.com/readthedocs/readthedocs.org/blob/d3351d52c41507f4c5f1ba737f8ef0b684d6e642/readthedocs/proxito/views/serve.py#L353-L353
   - https://github.com/readthedocs/readthedocs.org/blob/a59150c674fbb3bdc426183c623fe1ac37e64ec0/readthedocs/search/serializers.py#L137-L137

- version.is_sphinx_type

I have also wanted to get rid of using the doctype from some places (footer and search)

File: docs/user/build-customization.rst (hunk 11 above)
Comment (ericholscher) [reply docs/user/build-customization.rst:244]: Yea, the Pelican example is great. But I think we want something in addition, because otherwise it requires knowing what the pelican command means.

File: readthedocs/templates/builds/build_detail.html (hunk 17 above)
Comment (agjohnson) [reply readthedocs/templates/builds/build_detail.html:178]: To me, that doesn't feel like that's a case worth cluttering this UI for. There's any number of problems a user can introduce to their own configuration. Tuning this UI for edge case users, at the expense of normal usage, does negatively affect normal usage.

My main issue with this suggestion pattern in template logic is that multiple suggestions can show, regardless of build state, additional build errors, and existing notifications. And these suggestions are permanent, even after the user knows about it.

What about instead making this suggestion (but really all of these template suggestions eventually) notifications to the maintainer instead? This is triggered by first use of the feature or first project to use the feature, so we are providing the user with more obvious warning, which can be dismissed after reading, and we don't have to manage the notification with template logic.

File: readthedocs/doc_builder/director.py (hunk 8 above)
Comment (humitos) [reply readthedocs/doc_builder/director.py:365]: Yeah, 😞 . Santos asked the same in his first comment. It was just moved without changes in the logic.

File: docs/user/build-customization.rst (hunk 11 above)
Comment (humitos) [reply docs/user/build-customization.rst:244]: I don't have a strong preference for the tooling used here, but I'd like to make it clear in the example _the output folder_. If we hide everything in a `conf.py` file or `pelicanconf.py` with the required settings will be hard to understand what's expected/required from the user.

Comment (humitos) [reply docs/user/build-customization.rst:244]: If you have any suggestions that work better than the current one, I'll update the example.

File: docs/user/config-file/v2.rst (hunk 12 above)
Comment (humitos) [reply docs/user/config-file/v2.rst:466]: Yeah, sounds good to me!

File: docs/user/build-customization.rst (hunk 13 above)
Comment (humitos) [reply docs/user/build-customization.rst:241]: I liked the suggestion.

I changed it a little bit to mention "full control" (instead of "more control") and also to keep the link to the previous section. This will be useful when users arrives at this section directly without knowing that there is something simpler to execute custom commands.

File: docs/user/config-file/v2.rst (hunk 6 above)
Comment (humitos) [reply docs/user/config-file/v2.rst:489]: Yeah, I mentioned this in the notes of the PR.

> We should probably make build.tools not required since it's possible to generate HTML without any of these languages as shown in the first example where only echo command is used

Should we make `build.tools` _not required_ on this initial version? That will require some work on the config file because it implemented making `.tools` the required one instead of `.os` to decide between `build.image` and `build.os`.

File: docs/user/build-customization.rst (hunk 5 above)
Comment (humitos) [reply docs/user/build-customization.rst:258]: @agjohnson 
> and might be worth mentioning that asdf is used underneath everything too (you probably need to asdf reshim)

It felt weird to me to mention an implementation detail as part of the "usage of our platform". What happens if we stop using `asdf` in the future, or change it for a different tool?

@ericholscher mentioned about avoid adding magic and calling `asdf reshim` after each command and I tend to agree there but...

>> We could execute asdf reshim python after each of the build.commands, for now as well
> This seems like something we should avoid. The whole goal here is to remove magic, so let's not add it back in.

However, as a user, I'd expect that calling `pip install pelican` will immediately make `pelican` executable available in the PATH. Why users will need to know about `asdf` to make usage of `pelican`? Also, note that this is currently a limitation of `asdf` and `python` but it does not happen with `nodejs`. If you do `npm install -g something` that `something` is available in the PATH immediately.

You may think, "Yes, you are adding `-g` to the command to install it globally". However, the global option works differently on `asdf` than on regular `npm` because the command is run as an unprivileged user, and not installed under a global folder. `asdf` is doing some magic underneath to make this work.

I noted in the notes/decisions that `asdf` executes post hooks on particular command names, and `asdf reshim` was automatically executed when `pip` was called. I linked the issue and PR in the PR's description.

That said, I'd like to find a solution that does not involve users knowing anything about `asdf` because that's an implementation detail that they shouldn't care about.

File: readthedocs/doc_builder/director.py (hunk 2 above)
Comment (humitos) [reply readthedocs/doc_builder/director.py:351]: I'll try to add a `GENERIC` constant and see what happens. After an initial grep of `DOCUMENTATION_CHOICES` and `DOCTYPE_CHOICES` I didn't find it used in many places.

Also, grepping for `SPHINX` I didn't find anything that could conflict with `GENERIC`. The code pointed by @stsewd makes sense but I think it will be fine.

Besides, as @ericholscher thought it seems that usually, the default is `SPHINX` and the special cases are for `SPHINX_HTMLDIR`. So, I think it could work using a new doctype 👍🏼

File: readthedocs/templates/builds/build_detail.html (hunk 17 above)
Comment (humitos) [reply readthedocs/templates/builds/build_detail.html:178]: I agree with Eric here. I added this message on this page for that particular case and also to remind the users this feature is in beta _every_ time they take a look at the build page.

I prefer to not have a discussion about how to present notifications to users in this PR. I'd like to see a completely different approach than the one we currently have, but we are not there yet. I'm just following the pattern we already have for now --which I agree, it's not great.

> Tuning this UI for edge case users, at the expense of normal usage, does negatively affect normal usage.

I'm fine keeping the message, for now, we are launching a new beta feature to over-communicate our users first and remove/change it later when the feature is more established and people know more about it and its limitations. I don't think this will have a huge impact on many users.

File: docs/user/config-file/v2.rst (hunk 12 above)
Comment (humitos) [reply docs/user/config-file/v2.rst:466]: I went ahead and used `_readthedocs/html`

File: docs/user/build-customization.rst (hunk 5 above)
Comment (agjohnson) [reply docs/user/build-customization.rst]: Yeah I agree with that, I don't think it serves us well to surface `asdf` internals too much.

> This seems like something we should avoid. The whole goal here is to remove magic, so let's not add it back in.

I feel like we do need to define "magic" that we don't want, as pushing towards "no magic" isn't necessarily what we want. The flavor of magic we don't want is where we assume we know what a user meant to do and do it for them.

In this case, it's probably safe to say the user meant to have the binary available after installing the package, there isn't really another option.

But, I'd agree that a post pip install hook makes the most sense. Executing reshim liberally, after each command, might make it easier to give user some unwanted side effects.

File: readthedocs/templates/builds/build_detail.html (hunk 17 above)
Comment (ericholscher) [reply readthedocs/templates/builds/build_detail.html:178]: Strong agree on not trying to build a better notification system in this PR. I think this is warranted since it's a beta feature. This UX will be removed when we make it not beta, but for now the benefits outweigh the UX issues.

File: docs/user/build-customization.rst [hunk 18]
```diff
@@ -4,7 +4,13 @@ Build customization
...
 but we offer additional customization to support more uses of our platform.
 This page explains how to extend the build process using :term:`user-defined build jobs` to execute custom commands,
-and also how to :ref:`override the build process completely <build-customization:override the build process completely>`.
+and also how to override the build process completely:
+
+`Extend the build process`_
+    If you are using Sphinx or Mkdocs and need to execute additional commands.
+
+`Override the build process`_
+    If you want full control over your build or are not using Sphinx or Mkdocs.
```
Comment (ericholscher) [inline docs/user/build-customization.rst:13]:
```suggestion
    If you want full control over your build. This option supports any tool that generates HTML as part of the build.
```

File: docs/user/build-customization.rst [hunk 19]
```diff
@@ -225,23 +239,26 @@ To setup it, you need to define the version of Node.js to use and install the de
...
 
 .. warning::
 
    This feature is in a *beta phase* and could suffer incompatible changes or even removed completely in the near feature.
-   It does not yet support some of the Read the Docs' integrations like the :term:`flyout menu`, search and ads.
-   However, integrating all of them is part of the plan.
-   Use it under your own responsibility.
+   It does not yet support some of the Read the Docs' features like the :term:`flyout menu`, search and ads.
+   However, we do plan to support these features in the future.
+   Use this under your own risk.
```
Comment (ericholscher) [inline docs/user/build-customization.rst:250]:
```suggestion
   Use this feature at your own risk.
```

File: docs/user/config-file/v2.rst [hunk 20]
```diff
@@ -473,7 +473,7 @@ Specify a list of commands that Read the Docs will run on the build process.
 When ``build.commands`` is used, none of the :term:`pre-defined build jobs` will be executed.
 (see :doc:`/build-customization` for more details).
 This allows you to run custom commands and control the build process completely.
-The ``output/`` directory (relative to the checkout's path) will be uploaded and hosted by Read the Docs.
+The ``_readthedocs/html`` directory (relative to the checkout's path) will be uploaded and hosted by Read the Docs.
```
Comment (ericholscher) [inline docs/user/config-file/v2.rst:476]: This still feels really small compared to how important the information is, but we can improve it over time.

**Review by ericholscher — APPROVED** [review]
This looks close to me. I think we're close enough to get it out next week for testing. There's lots of additional work to be done in docs, etc. -- but this is enough for now.

File: readthedocs/templates/builds/build_detail.html (hunk 17 above)
Comment (agjohnson) [reply readthedocs/templates/builds/build_detail.html:178]: If you want to do the work in bulk, that's okay, but I also don't feel there is anything stopping us from using a modern pattern here. We've used per-view notifications in the past and I feel we discussed a similar build notification not too long ago even. Are we expecting the pattern to not use a known pattern like notifications?

But we do need to prioritize this work either way. These blocks make templates harder to manage and pose design issues. I didn't even bother implementing this logic/blocks in the new templates build detail page.

Comment (agjohnson) [reply readthedocs/templates/builds/build_detail.html:178]: > That said, I think the template notification is good enough for now for this use case and I'm happy to work on a full replacement when working on the notifications project.

Fair enough. Lets keep up pressure to solve #9279 early then, and I'd love to prioritize the (much larger) #3399.

My main concern is getting stuck in the cyclical pattern of using the template logic because it's there, and then not making progress on these two issues because we've added too much debt for the problem to be solved easily.

Any little thing we can do to break this cycle is important.

File: docs/user/build-customization.rst (hunk 5 above)
Comment (humitos) [reply docs/user/build-customization.rst]: `pyenv`, which is used by `asdf` to compile Python versions, is already calling `pyenv-reshim` when it detects `install` or `uninstall` in the command executed: https://github.com/pyenv/pyenv/blob/master/pyenv.d/exec/pip-rehash/pip#L26

(note they have the same for `conda`, `pip` and `easy_install` in the same directory https://github.com/pyenv/pyenv/tree/master/pyenv.d/exec/pip-rehash)

----

`asdf` implemented the pre/post hooks but they had to revert them because it introduced other problems due to the change to not use `exec` anymore: https://github.com/asdf-vm/asdf/issues/475

However, the `asdf-python` plugin has a PR open to wrap the pip command to reshim automatically as `asdf-nodejs` does. This is why it does work for nodejs' `npm` but it does not work for `pip` currently.

We want that PR merged, which will solve our problem. I'll follow up there with them.

Comment (humitos) [reply docs/user/build-customization.rst]: I went ahead and implemented option 2) for now. The example from `test-builds` without calling `asdf reshim` manually by the user now works out of the box 🎉 

Let me know if you want to change directions here and we can continue talking about this to find the best path at this point.

File: readthedocs/doc_builder/director.py (hunk 2 above)
Comment (humitos) [reply readthedocs/doc_builder/director.py]: I added a `GENERIC` type and I didn't find too many problems. Everything seems to continue working fine. I'm not sure if we will find some issues in the future because of this, but at least nothing obvious at this point.

As I said, I'm not really aware of how the doctype is used in the code, but I checked all the files I was pointed to and I didn't find anything that called my attention and I thought it could break at least.

Actually, while working on this I found a bug. We weren't updating the `Version.documentation_type` based on the `doctype` used in the config file --so, all the `Version.documentation_type` are defaulting to the `Project.documentation_type`. See this commit for the fix: [`99efb00` (#9150)](https://github.com/readthedocs/readthedocs.org/pull/9150/commits/99efb00bad5138d7d609b9474840c416dce3f560)

Let me know if there is something else here I should care about.

**Comment by humitos** [comment]
I updated the PR with all the suggestions that I received and improved the code a little bit. Now, it handles reshiming automatically when it detects a Python package manager installing something and it also uses a `'generic'` doctype for this type of documentation (`build.commands`).

For now, we are using template-based notifications but we plan to change that because none of us like this pattern. However, it's _good enough_, for now, to move forward and don't block this feature on that.

`build.tools` is required for now, but I opened an issue to track that and remove that restriction. I'm not doing it in this PR because it involves a good refactor of the `ConfigFile` object due that it differentiates legacy from the newer pattern based on `build.tools` instead of `build.os`.

File: docs/user/build-customization.rst [hunk 21]
```diff
@@ -218,3 +237,39 @@ To setup it, you need to define the version of Node.js to use and install the de
...
+   It does not yet support some of the Read the Docs' features like the :term:`flyout menu`, search and ads.
+   However, we do plan to support these features in the future.
+   Use this feature at your own risk.
+
+
+If your project requires full control of the build process,
+and :ref:`extending the build process <build-customization:extend the build process>` is not enough,
+all the commands executed during builds can be overridden using the :ref:`config-file/v2:build.commands` configuration file key.
+
+For example, if your project uses `Pelican <https://blog.getpelican.com/>`_ instead of Sphinx for it's documentation,
```
Comment (stsewd) [inline docs/user/build-customization.rst:257]:
```suggestion
For example, if your project uses `Pelican <https://blog.getpelican.com/>`_ instead of Sphinx for its documentation,
```

File: readthedocs/doc_builder/director.py [hunk 22]
```diff
@@ -335,6 +340,185 @@ def run_build_job(self, job):
...
+        for command in self.data.config.build.commands:
+            environment.run(*command.split(), escape_command=False, cwd=cwd)
+
+            # Execute ``asdf reshim python`` if the user is installing a
+            # package since the package may contain an executable
+            for reshim_command in reshim_commands:
+                # Convert tuple/list into set to check reshim command is a
+                # subset of the command itself. This is to find ``pip install``
+                # but also ``pip -v install`` and ``python -m pip install``
+                if set(reshim_command).issubset(command.split()):
```
Comment (stsewd) [inline readthedocs/doc_builder/director.py:363]: you could declare `reshim_commands` as sets, so you don't have to convert it to a set every time.

File: readthedocs/projects/tasks/builds.py [hunk 23]
```diff
@@ -480,7 +480,7 @@ def on_success(self, retval, task_id, args, kwargs):
                 api_v2.version(self.data.version.pk).patch(
                     {
                         "built": True,
-                        "documentation_type": self.data.version.documentation_type,
+                        "documentation_type": self.data.config.doctype,
```
Comment (stsewd) [inline readthedocs/projects/tasks/builds.py:483]: the previous logic was correct, the real doctype for mkdocs is known at build time

https://github.com/readthedocs/readthedocs.org/blob/9cf20d0e6b05f091a41ac4278622797c692089d8/readthedocs/doc_builder/director.py#L357-L357

File: docs/user/build-customization.rst (hunk 5 above)
Comment (agjohnson) [reply docs/user/build-customization.rst]: I like option 2 as well :+1:

File: readthedocs/doc_builder/exceptions.py [hunk 24]
```diff
@@ -36,6 +37,9 @@ class BuildUserError(BuildBaseException):
         "To resolve this error, double check your project configuration and installed "
         "dependencies are correct and have not changed recently."
     )
+    BUILD_COMMANDS_WITHOUT_OUTPUT = gettext_noop(
+        f'No "{BUILD_COMMANDS_OUTPUT_PATH_HTML}" folder was created after building the project.'
```
Comment (agjohnson) [inline readthedocs/doc_builder/exceptions.py:41]:
```suggestion
        f'No "{BUILD_COMMANDS_OUTPUT_PATH_HTML}" folder was created during this build.'
```

**Review by agjohnson — APPROVED** [review]
This is looking great. :+1: from me with the last bits of new feedback resolved
